addons/mail/models/mail_link_preview.py

```diff
@@ -38,25 +39,58 @@ def _create_from_message_and_notify(self, message):
             'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; rv:91.0) Gecko/20100101 Firefox/91.0'
         })
         link_preview_values = []
+        link_previews_by_url = {
+            preview.source_url: preview for preview in message.sudo().link_preview_ids
+        }
         for url in list(urls):
+            if url in link_previews_by_url:
+                preview = link_previews_by_url.pop(url)
+                if not preview.is_hidden:
+                    link_previews += preview
+                continue
             if preview := link_preview.get_link_preview_from_url(url, requests_session):
                 preview['message_id'] = message.id
                 link_preview_values.append(preview)
-            if len(link_preview_values) > 5:
+            if len(link_preview_values) + len(link_previews) > 5:
                 break
+        for unused_preview in link_previews_by_url.values():
+            unused_preview._unlink_and_notify()
         if link_preview_values:
-            link_previews = self.env['mail.link.preview'].create(link_preview_values)
+            link_previews += link_previews.create(link_preview_values)
+        if link_previews:
             self.env['bus.bus']._sendone(message._bus_notification_target(), 'mail.record/insert', {
                 'LinkPreview': link_previews._link_preview_format()
             })
 
+    def _hide_and_notify(self):
+        if not self:
+            return True
+        notifications = [
+            (
+                link_preview.message_id._bus_notification_target(),
+                'mail.record/insert', {
+                    'Message': {
+                        'linkPreviews': [('DELETE', {'id': link_preview.id})],
+                        'id': link_preview.message_id.id,
+                    }
+                }
+            ) for link_preview in self
+        ]
+        self.env['bus.bus']._sendmany(notifications)
+        self.is_hidden = True
```

Comment:
(minor) I would update the value before sending on the bus (sometimes the bus data will contain the value, and it can prevent from sending obsolete value)